(Feature) Remove spam tokens / Deactivate tokens refresh fix

Agupane opened the pull request:
Closes #1324 
Closes #1329 

## Description
- Removes spam tokens by setting `exclude_spam` param in `fetchTokenCurrenciesBalances`, now the tokens marked as spam in the backend are not fetched and the total balance is not affected by them
- The "force fetch assets" feature was still there, but it wasn't working because the list of tokens within the `safe state` wasn't checked on `updateSafe` reducer, I fixed that replacing `merge` with `mergeDeep `
- Another thing solved here was the `currencyBalance/currencyRate` that were removed in [this PR](https://github.com/gnosis/safe-react/pull/1301#discussion_r484917828) and was causing two bugs: 
the first one is that the user lost the feature to store his preferred currency and avoid using always USD and the second one was related with the currency balances being deleted every time the user went to the assets tabs on this line:
```
    const storedCurrencies = await loadCurrencyValues()
    const storedCurrency = storedCurrencies[safeAddress]
    if (!storedCurrency) {
      return batch(() => {
        dispatch(setCurrencyBalances(safeAddress, List([])))
        dispatch(setSelectedCurrency(safeAddress, AVAILABLE_CURRENCIES.USD))
        dispatch(setCurrencyRate(safeAddress, 1))
      })
    }
```
- Also while adding some types, I found that `SerializedSessionState` wasn't working so I fixed it, but seems that we aren't using it, I would suggest removing it.

## Update
Now I left just the code needed, there are no fixes related to [this](https://github.com/gnosis/safe-react/pull/1336) (CC @dasanra)

I also left only the `spam` parameter, but seems that some tokens like `Cunning` or `Scatter` are still there, as far as I discussed with Lukas, they should be considered spam, just let you know in case you need to update the db @Uxio0 


## Update 14/09:
Closes #1324 
Closes #1329 

This PR fixes the issue with reloading the tokens and also adds the spam parameter for removing spam tokens from the balances list.

- Also add types and replaces old class component to functional component

gnosis-info commented:
Travis automatic deployment:
 https://pr1331--safereact.review.gnosisdev.com/app

gnosis-info commented:
Travis automatic deployment:
 https://pr1331--safereact.review.gnosisdev.com/app

dasanra commented on src/logic/currencyValues/api/fetchTokenCurrenciesBalances.ts:
```diff
@@ -11,9 +11,13 @@ export type BalanceEndpoint = {
+const fetchTokenCurrenciesBalances = (
+  safeAddress: string,
+  excludeSpamTokens = true,
+  onlyTrustedTokens = true,
```
Should we add this parameters as environment configuration and activate just on mainnet?

I'm having a strange behaviour with Rinkeby tokens. With this options activated I see 0 balance but I still see the tokens on the list.

Also in Rinkeby most of the tokens are not trusted/validated that's why I ask to activate on mainnet environments

mikheevm replied:
@Uxio0 could you please give us some input on this?

Uxio0 replied:
Maybe is a bug, @dasanra @mikheevm could you provide the Safe you are using for testing?

dasanra commented on src/store/index.ts:
```diff
@@ -86,7 +89,7 @@ export type AppReduxState = CombinedState<{
   [COOKIES_REDUCER_ID]: Map<string, any>
   [ADDRESS_BOOK_REDUCER_ID]: AddressBookReducerMap
-  [CURRENT_SESSION_REDUCER_ID]: Map<string, any>
+  [CURRENT_SESSION_REDUCER_ID]: SerializedSessionState
```
In your comment you say we are not using it and we should remove it.
But here we are using it?

Agupane replied:
Yes but it's not consumed anywhere, it's just added there

mikheevm commented on src/logic/currencyValues/store/middleware/index.ts:
```diff
@@ -1,18 +1,38 @@
+
+        const currencyValuesWithoutBalances = currencyValues.map((entry: CurrencyReducerMap) => {
+          return {
+            currencyRate: entry.get('currencyRate'),
```
we had a discussion on why we don't need the rates saved in localStorage and they can be removed and it gets brought back... why?

Agupane replied:
I just restored the entire changes from the commit in which the bug was added. If you want to remove it, please open a pr with that change

mikheevm replied:
Done: https://github.com/gnosis/safe-react/pull/1333

mikheevm commented on src/logic/currentSession/store/reducer/currentSession.ts:
```diff
@@ -7,24 +6,23 @@ import { saveCurrentSessionToStorage } from 'src/logic/currentSession/utils'
     },
   },
-  Map(),
+  [],
```
So the whole current session is now just an array with addresses with visited safes? What happens if we want to add a new prop there?

Agupane replied:
We are not using that state, I would suggest removing it completely

mikheevm replied:
> We are not using that state

It's used in incoming transactions middleware

> What happens if we want to add a new prop there?

This does not really answer the question

Agupane replied:
If you wanna add a prop in the future, just do the change..

mikheevm replied:
And refactor other stuff that was dependent on the previous structure? It had a structure where you could easily add new properties and you're changing it. Your change doesn't make any sense and makes it harder for other developers to do any changes in the future. I agree that it would make sense if we wanted to ship the feature really hard but your change is making something bad that was done in a good way

mikheevm commented on src/logic/currentSession/store/reducer/currentSession.ts:
```diff
@@ -7,24 +6,23 @@ import { saveCurrentSessionToStorage } from 'src/logic/currentSession/utils'
-export type SerializedSessionState = {
-  viewedSafes: string[]
-}
+export type SerializedSessionState = string[]
```
What are we serializing there?

Agupane replied:
Same as before

mikheevm replied:
It's not serialized, previously we would prefix the type with serialized if the state would be an immutable.js js data structure and we would convert it to json

mikheevm commented on src/logic/safe/hooks/useFetchTokens.tsx:
```diff
@@ -10,11 +10,12 @@ import { fetchTokens } from 'src/logic/tokens/store/actions/fetchTokens'
 import { Dispatch } from 'src/logic/safe/store/actions/types.d'
 
-export const useFetchTokens = (safeAddress: string): void => {
+export const useFetchTokens = (safeAddress: string | undefined): void => {
```
why do we add undefined there if this function cannot work as intended if the address is undefined?

Agupane replied:
Because the `safeAddress selector` could return string or undefined and hooks cannot be called conditionally, the check condition must be done on the hook

mikheevm replied:
Ideally, this prop would come from top-level components to other ones, and somewhere on its way, we would perform a check whether its defined, so we render one set of components only if this address s defined. The selector is totally valid and the problem is how we're using it. Fixing it with `as` statement is ok because we know if that component is now rendered the address is defined.

Agupane replied:
Ok so you propose instead of fixing the bug, that was caused by a refactor, do another refactor for the entire app to avoid an if inside a hook? that doesn't make sense to me, I prefer to delivery the fix

mikheevm replied:
@Agupane 
> Fixing it with as statement is ok because we know if that component is now rendered the address is defined.

Agupane commented:
@Uxio0 please check the updated comment on the PR description
@dasanra I removed all the other things non related to the fix, you can review this now

gnosis-info commented:
Travis automatic deployment:
 https://pr1331--safereact.review.gnosisdev.com/app

Uxio0 commented:
> @Uxio0 please check the updated comment on the PR description

@lukasschor is the one now updating the tokens db :wink:

mikheevm commented on src/logic/tokens/store/actions/fetchTokens.ts:
```diff
@@ -91,7 +92,7 @@ export const getTokenInfos = async (tokenAddress: string): Promise<Token> => {
 }
 
-export const fetchTokens = () => async (dispatch, getState) => {
+export const fetchTokens = () => async (dispatch: Dispatch, getState: () => AppReduxState): Promise<void> => {
```
this should be of type `ThunkDispatch` from react-thunk page we're using

gnosis-info commented:
Travis automatic deployment:
 https://pr1331--safereact.review.gnosisdev.com/app

mikheevm commented:
Could you please update PR title and desc to what it really does?

I would drop fixes for #1312 in favor of https://github.com/gnosis/safe-react/pull/1333

Agupane commented:
It's done @mikheevm

mikheevm commented on src/routes/safe/components/Balances/Coins/index.tsx:
```diff
@@ -84,16 +73,16 @@ const Coins = (props: Props): React.ReactElement => {
-  useMemo(() => {
-    setFilteredData(getBalanceData(activeTokens, selectedCurrency, currencyValues, currencyRate))
-  }, [activeTokens, selectedCurrency, currencyValues, currencyRate])
+  const filteredData: List<BalanceData> = React.useMemo(
```
Nitpick: we use effect via import, but useMemo is used via accessing it on React namespace. Could we make it consistent?

gnosis-info commented:
Travis automatic deployment:
 https://pr1331--safereact.review.gnosisdev.com/app

mikheevm reviewed (approved):
How does it fix deactivated tokens appear again? Was it an issue with mergeDeep too?

Agupane commented:
> How does it fix deactivated tokens appear again? Was it an issue with mergeDeep too?

I thought that because the bug seems fixed but I tested with another safeAddress and the problem was still there, anyway, I fixed it now and refactored an old class component to functional component

gnosis-info commented:
Travis automatic deployment:
 https://pr1331--safereact.review.gnosisdev.com/app

mikheevm commented on src/logic/safe/store/actions/updateActiveTokens.ts:
```diff
@@ -1,4 +1,6 @@
 import updateSafe from './updateSafe'
+import { Set } from 'immutable'
+import { Dispatch } from 'redux'
```
We have already spoke about Dispatch and ThunkDispatch differences, haven't we?

fernandomg reviewed (approved):
works great! :100:

gnosis-info commented:
Travis automatic deployment:
 https://pr1331--safereact.review.gnosisdev.com/app
